// carbon-components-react/es/components/Tabs/next/Tabs.stories.js
/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import React from 'react';
import { Tab, Button, Tabs, unstable_ContainedTab as ContainedTab, unstable_ContainedTabs as ContainedTabs } from 'carbon-components-react';
import { unstable_FeatureFlags as FeatureFlags } from 'carbon-components-react';
export default {
  title: 'Components/Tabs',
  decorators: [function (Story) {
    return /*#__PURE__*/React.createElement(FeatureFlags, {
      flags: {
        'enable-v11-release': true
      }
    }, /*#__PURE__*/React.createElement(Story, null));
  }],
  parameters: {
    component: Tabs,
    subcomponents: {
      Tab: Tab
    }
  }
};
export var Default = function Default() {
  return /*#__PURE__*/React.createElement(Tabs, null, /*#__PURE__*/React.createElement(Tab, {
    id: "tab-1",
    label: "Tab label 1"
  }, /*#__PURE__*/React.createElement("p", null, "Content for first tab goes here.")), /*#__PURE__*/React.createElement(Tab, {
    id: "tab-2",
    label: "Tab label 2"
  }, /*#__PURE__*/React.createElement("p", null, "Content for second tab goes here."), /*#__PURE__*/React.createElement(Button, null, "With a button")), /*#__PURE__*/React.createElement(Tab, {
    id: "tab-3",
    label: "Tab label 3",
    disabled: true
  }, /*#__PURE__*/React.createElement("p", null, "Content for third tab goes here.")), /*#__PURE__*/React.createElement(Tab, {
    id: "tab-4",
    label: "Tab label 4 shows truncation",
    title: "Tab label 4 shows truncation"
  }, /*#__PURE__*/React.createElement("p", null, "Content for fourth tab goes here.")), /*#__PURE__*/React.createElement(Tab, {
    label: /*#__PURE__*/React.createElement("div", null, "Custom Label")
  }, /*#__PURE__*/React.createElement("p", null, "Content for fifth tab goes here.")));
};
export var Contained = function Contained() {
  return /*#__PURE__*/React.createElement(ContainedTabs, null, /*#__PURE__*/React.createElement(ContainedTab, {
    id: "tab-1",
    label: "Tab label 1"
  }, /*#__PURE__*/React.createElement("p", null, "Content for first tab goes here.")), /*#__PURE__*/React.createElement(ContainedTab, {
    id: "tab-2",
    label: "Tab label 2"
  }, /*#__PURE__*/React.createElement("p", null, "Content for second tab goes here."), /*#__PURE__*/React.createElement(Button, null, "With a button")), /*#__PURE__*/React.createElement(ContainedTab, {
    id: "tab-3",
    label: "Tab label 3",
    disabled: true
  }, /*#__PURE__*/React.createElement("p", null, "Content for third tab goes here.")), /*#__PURE__*/React.createElement(ContainedTab, {
    id: "tab-4",
    label: "Tab label 4 shows truncation",
    title: "Tab label 4 shows truncation"
  }, /*#__PURE__*/React.createElement("p", null, "Content for fourth tab goes here.")), /*#__PURE__*/React.createElement(ContainedTab, {
    label: /*#__PURE__*/React.createElement("div", null, "Custom Label")
  }, /*#__PURE__*/React.createElement("p", null, "Content for fifth tab goes here.")));
};
