// carbon-components-react/lib/components/Tabs/next/Tabs.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Contained = exports.Default = exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _carbonComponentsReact = require("carbon-components-react");

/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
var _default = {
  title: 'Components/Tabs',
  decorators: [function (Story) {
    return /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.unstable_FeatureFlags, {
      flags: {
        'enable-v11-release': true
      }
    }, /*#__PURE__*/_react.default.createElement(Story, null));
  }],
  parameters: {
    component: _carbonComponentsReact.Tabs,
    subcomponents: {
      Tab: _carbonComponentsReact.Tab
    }
  }
};
exports.default = _default;

var Default = function Default() {
  return /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.Tabs, null, /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.Tab, {
    id: "tab-1",
    label: "Tab label 1"
  }, /*#__PURE__*/_react.default.createElement("p", null, "Content for first tab goes here.")), /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.Tab, {
    id: "tab-2",
    label: "Tab label 2"
  }, /*#__PURE__*/_react.default.createElement("p", null, "Content for second tab goes here."), /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.Button, null, "With a button")), /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.Tab, {
    id: "tab-3",
    label: "Tab label 3",
    disabled: true
  }, /*#__PURE__*/_react.default.createElement("p", null, "Content for third tab goes here.")), /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.Tab, {
    id: "tab-4",
    label: "Tab label 4 shows truncation",
    title: "Tab label 4 shows truncation"
  }, /*#__PURE__*/_react.default.createElement("p", null, "Content for fourth tab goes here.")), /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.Tab, {
    label: /*#__PURE__*/_react.default.createElement("div", null, "Custom Label")
  }, /*#__PURE__*/_react.default.createElement("p", null, "Content for fifth tab goes here.")));
};

exports.Default = Default;

var Contained = function Contained() {
  return /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.unstable_ContainedTabs, null, /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.unstable_ContainedTab, {
    id: "tab-1",
    label: "Tab label 1"
  }, /*#__PURE__*/_react.default.createElement("p", null, "Content for first tab goes here.")), /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.unstable_ContainedTab, {
    id: "tab-2",
    label: "Tab label 2"
  }, /*#__PURE__*/_react.default.createElement("p", null, "Content for second tab goes here."), /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.Button, null, "With a button")), /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.unstable_ContainedTab, {
    id: "tab-3",
    label: "Tab label 3",
    disabled: true
  }, /*#__PURE__*/_react.default.createElement("p", null, "Content for third tab goes here.")), /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.unstable_ContainedTab, {
    id: "tab-4",
    label: "Tab label 4 shows truncation",
    title: "Tab label 4 shows truncation"
  }, /*#__PURE__*/_react.default.createElement("p", null, "Content for fourth tab goes here.")), /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.unstable_ContainedTab, {
    label: /*#__PURE__*/_react.default.createElement("div", null, "Custom Label")
  }, /*#__PURE__*/_react.default.createElement("p", null, "Content for fifth tab goes here.")));
};

exports.Contained = Contained;
